src/OAuth/ResourceOwner/TelegramResourceOwner.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+/*
+ * This file is part of the HWIOAuthBundle package.
+ *
+ * (c) Hardware Info <opensource@hardware.info>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace HWI\Bundle\OAuthBundle\OAuth\ResourceOwner;
+
+use HWI\Bundle\OAuthBundle\OAuth\Response\PathUserResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\LazyResponseException;
+
+/**
+ * @author zorn-v
+ */
+class TelegramResourceOwner extends GenericOAuth2ResourceOwner
+{
+    public const TYPE = 'telegram';
+
+    protected array $paths = [
+        'identifier' => 'id',
+        'nickname' => 'username',
+        'realname' => 'first_name',
+        'firstname' => 'first_name',
+        'lastname' => 'last_name',
+        'profilepicture' => 'photo_url',
+    ];
+
+    public function getAuthorizationUrl($redirectUri, array $extraParameters = [])
+    {
+        list($botId) = explode(':', $this->options['client_secret']);
+        $parameters = array_merge([
+            'bot_id' => $botId,
+            'origin' => $redirectUri,
+            'return_to' => $redirectUri,
+        ], $extraParameters);
+
+        return $this->normalizeUrl($this->options['authorization_url'], $parameters);
+    }
+
+    public function handles(Request $request)
+    {
+        if (!$request->query->has('code')) {
+            $js = sprintf('<script>location.href = "?code=" + new URLSearchParams(location.hash.substring(1)).get("tgAuthResult")</script>');
```

Comment:
This smells to me, why do we need to use a JavaScript hack for the redirect?

Reply:
Telegram redirects to `<return_to_url>#tgAuthResult=base64encodedData` so we can not get it from server.

Reply:
After you confirm login on telegram side there is redirect to url specified in `return_to` parameter (as in other resource owners), but redirect not with query params (like `callback?code=and&params=from&resource=owner`), but with `hash` (like `callback#tgAuthResult`). And we need somehow to pass it to server.

Official docs (https://core.telegram.org/widgets/login) desctibes only login via js widget, but I picked it out proper url (`popup_url`) from widget js https://telegram.org/js/telegram-widget.js?22

> Cause I don't see anything on the page you pointed in the docs.

You need telegram installed and then that page open bot for creating bots in there.